module-bsp/board/linux/hal/battery_charger/BatteryCharger.cpp

```diff
@@ -40,116 +27,117 @@ namespace hal::battery
+        notificationChannel = nullptr;
     }
 
-    void BatteryCharger::deinit()
+    AbstractBatteryCharger::Voltage BatteryCharger::getBatteryVoltage() const
     {
-        shouldRun      = false;
-        IRQQueueHandle = nullptr;
-        DCDQueueHandle = nullptr;
+        return dummyBatteryVoltageLevel;
     }
 
-    void BatteryCharger::processStateChangeNotification(std::uint8_t notification)
+    AbstractBatteryCharger::SOC BatteryCharger::getSOC() const
     {
-        if (isPlugged) {
-            if (batteryLevel > fullBattery) {
-                if (isPlugged && Store::Battery::get().level == fullBattery) {
-                    Store::Battery::modify().state = Store::Battery::State::ChargingDone;
-                }
-            }
-            else {
-                Store::Battery::modify().state = Store::Battery::State::Charging;
-            }
+        return batteryLevel;
+    }
+    AbstractBatteryCharger::ChargingStatus BatteryCharger::getChargingStatus() const
+    {
+        if (isPlugged && batteryLevel >= 100) {
+            return ChargingStatus::ChargingDone;
+        }
+        else if (isPlugged && batteryLevel < 100) {
+            return ChargingStatus::Charging;
         }
         else {
-            Store::Battery::modify().state = Store::Battery::State::Discharging;
+            return ChargingStatus::Discharging;
         }
-        batteryLevel                   = std::clamp(batteryLevel, unsigned{0}, fullBattery);
-        Store::Battery::modify().level = batteryLevel;
-        eventsHandler.onStatusChanged();
     }
 
-    int BatteryCharger::getBatteryVoltage()
+    void BatteryCharger::worker()
     {
-        return dummyBatteryVoltageLevel;
+        mkfifo(batteryFIFO, fifoFileAccessRights);
+
+        // Open FIFO for write only
+        int fd = open(batteryFIFO, O_RDONLY | O_NONBLOCK);
+
+        while (shouldRun) {
+            std::uint8_t buff[fifoBuffSize];
+            std::int32_t readBytes = read(fd, buff, fifoBuffSize);
+
+            if (readBytes > 0) {
+                Events evt{};
+                switch (static_cast<char>(buff[0])) {
+                case chargerPlugStateChange:
+                    isPlugged = !isPlugged;
+                    evt       = Events::Charger;
+                    break;
+                case batteryLevelUp:
+                    batteryLevel++;
+                    evt = Events::SOC;
+                    break;
+                case batteryLevelDown:
+                    batteryLevel--;
+                    evt = Events::SOC;
+                    break;
+                case chargerTypeDcdSDP:
+                case chargerTypeDcdCDP:
+                case chargerTypeDcdDCP:
+                    evt = Events::Charger;
+                    break;
+                default:
+                    continue;
+                }
+                xQueueSend(notificationChannel, &evt, queueTimeoutTicks);
+            }
+            vTaskDelay(taskDelay);
```

Comment:
can't we just hang on read? ( is sleep really needed?)

Reply:
The answer is no if we use a simple FreeRTOS queue mechanism, This is because we won't be able to handle the stop 'signal' invoked from BatteryCharger destructor. It does really matter that much for the simulator tho. Another solution is to use a more sophisticated messages system, for instance, `WorkQueue` used by rt1051.

Reply:
It is also possible to just invoke `vTaskDelete` from the destructor but I would consider it rather a rough way of exiting.